core/dnsserver/quic.go

```diff
@@ -0,0 +1,47 @@
+package dnsserver
+
+import (
+	"encoding/binary"
+	"github.com/miekg/dns"
+	"github.com/quic-go/quic-go"
+	"net"
+)
+
+type DoQWriter struct {
+	localAddr  net.Addr
+	remoteAddr net.Addr
+	stream     quic.Stream
+	Msg        *dns.Msg
+}
+
+func (w *DoQWriter) Write(b []byte) (int, error) {
+	b = AddPrefix(b)
+	return w.stream.Write(b)
+}
+
+func (w *DoQWriter) WriteMsg(m *dns.Msg) error {
+	bytes, err := m.Pack()
+	if err != nil {
+		return err
+	}
+
+	_, err = w.Write(bytes)
+	return err
+}
+
+// AddPrefix adds a 2-byte prefix with the DNS message length.
+func AddPrefix(b []byte) (m []byte) {
+	m = make([]byte, 2+len(b))
+	binary.BigEndian.PutUint16(m, uint16(len(b)))
+	copy(m[2:], b)
+
+	return m
+}
+
+// These methods implement the dns.ResponseWriter interface from Go DNS.
+func (w *DoQWriter) Close() error          { return nil }
```

Comment:
shouldnt you close the stream here?

Reply:
Makes sense. I moved the `STREAM FIN` sending mechanism to this method. I noticed it is not called in the `serveDNS` routine, so I explicitly called it in `w.WriteMsg`.